helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedInstanceCapacity.java

```diff
@@ -0,0 +1,164 @@
+import org.apache.helix.controller.dataproviders.ResourceControllerDataProvider;
+import org.apache.helix.controller.rebalancer.util.WagedValidationUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class WagedInstanceCapacity implements InstanceCapacityDataProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(WagedInstanceCapacity.class);
+
+  // Available Capacity per Instance
+  private Map<String, Map<String, Integer>> _perInstance;
+  private ResourceControllerDataProvider _cache;
+
+  public WagedInstanceCapacity(ResourceControllerDataProvider clusterData) {
+    _cache = clusterData;
+    _perInstance = new HashMap<>();
+
+    ClusterConfig clusterConfig = _cache.getClusterConfig();
+    for (InstanceConfig instanceConfig : _cache.getInstanceConfigMap().values()) {
+      Map<String, Integer> instanceCapacity =
+        WagedValidationUtil.validateAndGetInstanceCapacity(clusterConfig, instanceConfig);
+      _perInstance.put(instanceConfig.getInstanceName(), instanceCapacity);
+    }
+  }
+
+  /**
+   * Create Default Capacity Map.
+   * This is a utility method to create a default capacity map matching instance capacity map for participants.
+   * This is required as non-WAGED partitions will be placed on same instance and we don't know their actual capacity.
+   * This will generate default values of 0 for all the capacity keys.
+   */
+  public Map<String, Integer> createDefaultParticipantWeight() {
+    // copy the value of first Instance capacity.
+    Map<String, Integer> partCapacity = new HashMap<>(_perInstance.values().iterator().next());
+
+    // reset the value of all capacity to 0.
+    for (String key : partCapacity.keySet()) {
+      partCapacity.put(key, 0);
+    }
+    return partCapacity;
+  }
+
+  /**
+   * Process the pending messages based on the Current states
+   * @param currentState - Current state of the resources.
+   */
+  public void processPendingMessages(CurrentStateOutput currentState) {
+    Map<String, Map<Partition, Map<String, Message>>> pendingMsgs = currentState.getPendingMessages();
+    
+    for (String resource : pendingMsgs.keySet()) {
+      Map<Partition, Map<String, Message>> partitionMsgs = pendingMsgs.get(resource);
+
+      for (Partition partition : partitionMsgs.keySet()) {
+        String partitionName = partition.getPartitionName();
+
+        // Get Partition Weight
+        Map<String, Integer> partCapacity = getPartitionCapacity(resource, partitionName);
+
+        // TODO - check
+        Map<String, Message> msgs = partitionMsgs.get(partition);
+        // TODO - Check
+        for (String instance : msgs.keySet()) {
+           reduceAvailableInstanceCapacity(instance, partCapacity);
+        }
+      }
+    }
+  }
+
+  /**
+   * Get the partition capacity given Resource and Partition name.
+   */
+  private Map<String, Integer> getPartitionCapacity(String resource, String partition) {
```

Comment:
This should be in a separate class, the resource capacity provider, right?
FYI, https://github.com/apache/helix/pull/2480/files#diff-07032ec7fdb5667dcf36e472c77dfefdd1eb2cff72feb472d2158b1df6fd9664

Reply:
we can.

Reply:
will do this as part of change-2.